oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java

```diff
@@ -0,0 +1,449 @@
+  public static ExternalAccountCredentials fromJson(
+      Map<String, Object> json, HttpTransportFactory transportFactory) {
+    checkNotNull(json);
+    checkNotNull(transportFactory);
+
+    String audience = (String) json.get("audience");
+    String subjectTokenType = (String) json.get("subject_token_type");
+    String tokenUrl = (String) json.get("token_url");
+    String tokenInfoUrl = (String) json.get("token_info_url");
+    String serviceAccountImpersonationUrl = (String) json.get("service_account_impersonation_url");
+
+    Map<String, Object> credentialSourceMap = (Map<String, Object>) json.get("credential_source");
+
+    // Optional params.
+    String clientId = json.containsKey("client_id") ? (String) json.get("client_id") : null;
+    String clientSecret =
+        json.containsKey("client_secret") ? (String) json.get("client_secret") : null;
+    String quotaProjectId =
+        json.containsKey("quota_project_id") ? (String) json.get("quota_project_id") : null;
+
+    if (isAwsCredential(credentialSourceMap)) {
+      // TODO(lsirac) return AWS credential here.
+    }
+    return new IdentityPoolCredentials(
+        transportFactory,
+        audience,
+        subjectTokenType,
+        tokenUrl,
+        tokenInfoUrl,
+        new IdentityPoolCredentialSource(credentialSourceMap),
+        serviceAccountImpersonationUrl,
+        quotaProjectId,
+        clientId,
+        clientSecret,
+        /* scopes= */ null);
+  }
+
+  private static boolean isAwsCredential(Map<String, Object> credentialSource) {
+    return credentialSource.containsKey("environment_id")
+        && ((String) credentialSource.get("environment_id")).startsWith("aws");
+  }
+
+  /**
+   * Exchanges the 3PI credential for a GCP access token.
+   *
+   * @param stsTokenExchangeRequest the STS token exchange request.
+   * @return the access token returned by STS.
+   * @throws OAuthException if the call to STS fails.
+   */
+  protected AccessToken exchange3PICredentialForAccessToken(
+      StsTokenExchangeRequest stsTokenExchangeRequest) throws IOException {
+
+    StsRequestHandler requestHandler =
+        StsRequestHandler.newBuilder(
+                tokenUrl, stsTokenExchangeRequest, transportFactory.create().createRequestFactory())
+            .setHeaders(new HttpHeaders().setContentType("application/x-www-form-urlencoded"))
+            .build();
+
+    StsTokenExchangeResponse response = requestHandler.exchangeToken();
+    return response.getAccessToken();
+  }
+
+  /**
+   * Attempts service account impersonation.
+   *
+   * @param accessToken the access token to be included in the request.
+   * @return the access token returned by the generateAccessToken call.
+   * @throws IOException if the service account impersonation call fails.
+   */
+  protected AccessToken attemptServiceAccountImpersonation(AccessToken accessToken)
+      throws IOException {
+    if (serviceAccountImpersonationUrl == null) {
+      return accessToken;
+    }
```

Comment:
Would a caller of this need to know if this was skipped?

Reply:
What do you mean? Are you suggesting I log something here or?

Reply:
This is an internal method so it probably doesn't actually matter.

I was just wondering if somewhere that calls this method might want to conditionally do something else if we did not actually attempt impersonation.